src/common/server.ts

```diff
@@ -27,59 +27,90 @@ import {
 import { updateStatus } from "./status";
 import { getProjectRoot } from "./utilities";
 import { isVirtualWorkspace } from "./vscodeapi";
+import { exec } from "child_process";
+import which = require("which");
 
-export type IInitOptions = {
+export type IInitializationOptions = {
   settings: ISettings[];
   globalSettings: ISettings;
 };
 
-function createNativeServerOptions(settings: ISettings): Executable {
+// Function to execute a command and capture the stdout.
+function executeCommand(command: string): Promise<string> {
+  return new Promise((resolve, reject) => {
+    exec(command, (error, stdout, _) => {
+      if (error) {
+        reject(error);
+      } else {
+        resolve(stdout);
+      }
+    });
+  });
+}
+
+async function findRuffBinaryPath(settings: ISettings): Promise<string> {
   // 'path' setting takes priority over everything.
   if (settings.path.length > 0) {
     for (const path of settings.path) {
-      if (fsapi.existsSync(path)) {
+      if (await fsapi.pathExists(path)) {
```

Comment:
Could it be more precise and use `fsapi.ensureFile`?

Reply:
I don't think we can use that because as per the documentation it'll create the file and parent directories if it doesn't exists. What we want is to just check if it exists or not.

```ts
/**
 * Ensures that the file exists. If the file that is requested to be created is in
 * directories that do not exist, these directories are created. If the file already
 * exists, it is **NOT MODIFIED**.
 *
 * @example
 * import * as fs from 'fs-extra'
 *
 * const file = '/tmp/this/path/does/not/exist/file.txt'
 *
 * // With a callback:
 * fs.ensureFile(file, err => {
 *   console.log(err) // => null
 *   // file has now been created, including the directory it is to be placed in
 * })
 *
 * // With Promises:
 * fs.ensureFile(file)
 *  .then(() => {
 *    console.log('success!')
 *  })
 *  .catch(err => {
 *    console.error(err)
 *  })
 *
 * // With async/await:
 * async function asyncAwait () {
 *   try {
 *     await fs.ensureFile(file)
 *     console.log('success!')
 *   } catch (err) {
 *     console.error(err)
 *   }
 * }
 *
 * asyncAwait()
 */
```